components/raftstore/src/store/peer.rs

```diff
@@ -2067,6 +2067,12 @@ where
                 false
             }
             Ok(Either::Left(idx)) => {
+                if self.has_applied_to_current_term() {
+                    // After this peer has applied to current term and passed above checking including `cmd_epoch_checker`,
+                    // we can safely guarantee that this proposal will be committed if there is no abnormal leader transfer
+                    // in the near future. Thus proposed callback can be called.
+                    cb.invoke_proposed();
```

Comment:
Does it mean that the callback is possibly not called when proposing is finished?

Reply:
Yes. But `write_cb` will be called definitely.

Reply:
@youjiali1995 I'm afraid that if it's *possibly* (instead of *always*) called on proposed, it's not very match to the name "proposed_cb" semantically. Perhaps it's better to pass some context to the callback (for example, pass the condition `has_applied_to_current_term` to it). It's not very important though.

Reply:
I think maybe all proposals should be rejected if the peer has not applied to current term. We do the write after getting a snapshot which almost ensures that.
The only different case as below:
1. After getting a snapshot, the leader can not connect to others and steps back to followers.
2. Then it becames leader again. 
In that case, the term is one more than before. If this leader has not applied to current term, the write proposal will be rejected while in previous code it can be accepted but `proposed_cb` is not called.
IMO this case is very unlikely to happen thus can not affect performance so maybe we should do this thing and make logic more clear.

Reply:
Should we also check `is_splitting` and `is_merging` here?

Reply:
But we have check term in another place before.